cmd/contour/serve.go

```diff
@@ -187,6 +187,10 @@ func validateCRDs(dynamicClient dynamic.Interface, log logrus.FieldLogger) {
 
 // doServe runs the contour serve subcommand.
 func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	ectx := context.Background()
+	serveContext, cancel := context.WithCancel(ectx)
+	defer cancel()
```

Comment:
I don't think you need to manage cancellation explicitly here, since the work group should already do that for us; I think we just need to switch a few of the tasks to use `g.AddContext(...)` instead of `g.Add(...)`

Reply:
Workgroup needs a way to get the sigterm and shutdown which is why I had this calling `cancel()` when it received that signal. Let me see if I can put that into the workgroup and remove the sigterm bits from serve.go all together.

Reply:
You should be able to leave the sigterm handling as-is; when it receives a sigterm, that function returns, which then causes the workgroup to signal all the other tasks to stop. But that requires all the other tasks to get their cancellation from the workgroup too.